src/TraceEvent/EventPipe/NetTraceFormat_v6.md

```diff
@@ -0,0 +1,288 @@
+- BeginPrivateObject Tag
+- SERIALIZED TYPE
+  - BeginPrivateObject Tag
+  - NullReference Tag
+  - version (int)
+  - minimum reader version (int)
+  - type name (string)
+  - EndPrivateObject Tag
+
+This header is 16 bytes + a variable length UTF8 string indicating the actual type of the object. The length of the object content was not included in the header. The EventBlock, MetadataBlock, and SPBlock objects included a BlockSize field as the first 4 bytes of the object content and the Trace object had fixed size.
+
+In V6 the object header is replaced by a simplified block header:
+
+```
+struct BlockHeader
+{
+  uint24 BlockSize
+  uint8 BlockKind
+}
+```
+
+Readers may want to read the BlockHeader as a single little endian 4 byte integer X where BlockSize = X & 0xFFFFFF and BlockKind = X >> 24.
+
+
+The type name strings used in the object header type name field map to new BlockKind values:
+
+```
+enum BlockKind
+{
+  EndOfStream = 0 // this is a new block type described later
+  Trace = 1,
+  Event = 2,
+  Metadata = 3,
+  Stack = 4,
+  SequencePoint = 5
+}
+```
+
+In V6 the EventBlock, MetadataBlock, StackBlock and SPBlock no longer have the initial BlockSize and padding fields. The value previously stored in the BlockSize object field is now in the BlockHeader. The semantics of the BlockSize header field are very similar to before, it represents the number of bytes in the remainder of the block. Specifically if file offset X points to the start of a block, then X + sizeof(BlockHeader) + BlockSize points to the start of the next block.
+
+Readers should skip over any blocks with an unrecognized BlockKind. New blocks might be introduced in minor version updates to store new types of data and this is expected to be a back-compatible change because old readers skip over it.
+
+
+#### Removed block footer
+
+In V5 all objects ended with a trailing EndPrivateObject tag byte. In V6 this trailing byte serves no purpose and is removed.
+
+#### Changed end of stream marker
+
+In V5 the file ended with a NullReferenceTag byte. In V6 the file ends with a new EndOfStream block. This block consists only of a BlockHeader and has BlockSize=0.
+
+## Multi-process support
+
+### Motivation
+
+Up to V5 the format was always designed to trace a single process only. Now we'd like to encode multi-process traces as well. This requires including process id information in places that previously only included thread ids. We also need to adjust definitions of a few TraceBlock fields.
+
+### Changes
+
+#### New ThreadBlock
+
+A new top-level block, BlockHeader.Kind=6, contains a set of thread id/process id tuples referenced by index. This allows many events to refer the same (thread id, process id) without needing to repeatedly encode the same pair of potentially large IDs.
+
+The content of a thread block is:
+
+- Count - uint16
+- Concatenated sequence of Count entries, each of which is
+  - Index - varuint
+  - ThreadId - varuint
```

Comment:
In EventPipe the thread id is a uint64_t sized value, but currently our TraceEvent handles it as int, https://github.com/microsoft/perfview/issues/1446. This has caused issues with negative thread id's showing up in perfview, there are other platforms that typedef pthread_t to a pointer (iOS/tvOS/MacCatalyst/MacOS), so thread id will be truncated on these platforms as well. If we change the file format, I also believe we should try to enhance libraries parsing this data to either switch to a wider data type for thread id, or utilize the index as an indirection to get to the full thread id and process id.

Reply:
Is this a suggestion for a change to the format, or a suggestion to change TraceEvent? From what I can tell both the current and new format will happily support a 64-bit ThreadId. The parser for NetTrace also supports reading/storing a 64 bit thread in its own internal datastructure. The problem arises because the interface between the NetTrace parser and the rest of TraceEvent forces the parser to downcast the thread ID to a 32 bit value. Fixing that is the work tracked by #1446. I couldn't tell that anything we do to change the format or the NetTrace parser changes the work involved to solve #1446 ?

Reply:
This was mainly just a reminder of the issue in the TraceEvent library and if we implement a new stuff to parse format, we should consider adding a new API to handle thread id as a 64-bit unisgned data type all the way through the public API. Its not directly related to the format, since we already handle it as 64-bit in the format its not directly related to this issue. I make a note in #1446.